File: ios/Sources/ExamplePlugin/ExamplePlugin.swift
import Foundation
import Capacitor

/**
 * Zyprint plugin for iOS
 * Capacitor iOS Plugin Development Guide: https://capacitorjs.com/docs/plugins/ios
 */
@objc(ZyprintPlugin)
public class ZyprintPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ZyprintPlugin"
    public let jsName = "Zyprint"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "echo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "discoverPrinters", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "connectToPrinter", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "disconnectFromPrinter", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "printText", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "printReceipt", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPrinterStatus", returnType: CAPPluginReturnPromise)
    ]
    
    private var implementation = Zyprint()
    
    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        let result = implementation.echo(value)
        call.resolve([
            "value": result
        ])
    }
    
    @objc func discoverPrinters(_ call: CAPPluginCall) {
        implementation.discoverPrinters { printers in
            let printersData = printers.map { printer in
                return [
                    "identifier": printer.identifier,
                    "model": printer.model,
                    "status": printer.status
                ]
            }
            
            call.resolve([
                "printers": printersData
            ])
        }
    }
    
    @objc func connectToPrinter(_ call: CAPPluginCall) {
        guard let identifier = call.getString("identifier") else {
            call.reject("Missing identifier parameter")
            return
        }
        
        implementation.connectToPrinter(identifier: identifier) { success, error in
            if success {
                call.resolve([
                    "connected": true
                ])
            } else {
                call.reject(error ?? "Connection failed")
            }
        }
    }
    
    @objc func disconnectFromPrinter(_ call: CAPPluginCall) {
        guard let identifier = call.getString("identifier") else {
            call.reject("Missing identifier parameter")
            return
        }
        
        implementation.disconnectFromPrinter(identifier: identifier) { success in
            call.resolve([
                "disconnected": success
            ])
        }
    }
    
    @objc func printText(_ call: CAPPluginCall) {
        guard let text = call.getString("text"),
              let identifier = call.getString("identifier") else {
            call.reject("Missing required parameters")
            return
        }
        
        implementation.printText(text, identifier: identifier) { success, error in
            if success {
                call.resolve([
                    "success": true
                ])
            } else {
                call.reject(error ?? "Print failed")
            }
        }
    }
    
    @objc func printReceipt(_ call: CAPPluginCall) {
        guard let template = call.getObject("template"),
              let identifier = call.getString("identifier") else {
            call.reject("Missing required parameters")
            return
        }
        
        implementation.printReceipt(template, identifier: identifier) { success, error in
            if success {
                call.resolve([
                    "success": true
                ])
            } else {
                call.reject(error ?? "Print failed")
            }
        }
    }
    
    @objc func getPrinterStatus(_ call: CAPPluginCall) {
        guard let identifier = call.getString("identifier") else {
            call.reject("Missing identifier parameter")
            return
        }
        
        implementation.getPrinterStatus(identifier: identifier) { status, paperStatus, connected in
            call.resolve([
                "status": status,
                "paperStatus": paperStatus,
                "connected": connected
            ])
        }
    }
}

File: android/src/main/java/com/mycompany/plugins/example/ExamplePlugin.java
package com.mycompany.plugins.example;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "Zyprint")
public class ZyprintPlugin extends Plugin {

    private Zyprint implementation = new Zyprint();

    @PluginMethod
    public void echo(PluginCall call) {
        String value = call.getString("value");

        JSObject ret = new JSObject();
        ret.put("value", implementation.echo(value));
        call.resolve(ret);
    }

    @PluginMethod
    public void discoverPrinters(PluginCall call) {
        implementation.discoverPrinters(new Zyprint.PrinterDiscoveryCallback() {
            @Override
            public void onPrintersFound(JSArray printers) {
                JSObject ret = new JSObject();
                ret.put("printers", printers);
                call.resolve(ret);
            }

            @Override
            public void onError(String error) {
                call.reject(error);
            }
        });
    }

    @PluginMethod
    public void connectToPrinter(PluginCall call) {
        String identifier = call.getString("identifier");
        if (identifier == null) {
            call.reject("Missing identifier parameter");
            return;
        }

        implementation.connectToPrinter(identifier, new Zyprint.ConnectionCallback() {
            @Override
            public void onConnected() {
                JSObject ret = new JSObject();
                ret.put("connected", true);
                call.resolve(ret);
            }

            @Override
            public void onError(String error) {
                call.reject(error);
            }
        });
    }

    @PluginMethod
    public void disconnectFromPrinter(PluginCall call) {
        String identifier = call.getString("identifier");
        if (identifier == null) {
            call.reject("Missing identifier parameter");
            return;
        }

        implementation.disconnectFromPrinter(identifier, new Zyprint.DisconnectionCallback() {
            @Override
            public void onDisconnected() {
                JSObject ret = new JSObject();
                ret.put("disconnected", true);
                call.resolve(ret);
            }

            @Override
            public void onError(String error) {
                call.reject(error);
            }
        });
    }

    @PluginMethod
    public void printText(PluginCall call) {
        String text = call.getString("text");
        String identifier = call.getString("identifier");
        
        if (text == null || identifier == null) {
            call.reject("Missing required parameters");
            return;
        }

        implementation.printText(text, identifier, new Zyprint.PrintCallback() {
            @Override
            public void onSuccess() {
                JSObject ret = new JSObject();
                ret.put("success", true);
                call.resolve(ret);
            }

            @Override
            public void onError(String error) {
                call.reject(error);
            }
        });
    }

    @PluginMethod
    public void printReceipt(PluginCall call) {
        JSObject template = call.getObject("template");
        String identifier = call.getString("identifier");
        
        if (template == null || identifier == null) {
            call.reject("Missing required parameters");
            return;
        }

        implementation.printReceipt(template, identifier, new Zyprint.PrintCallback() {
            @Override
            public void onSuccess() {
                JSObject ret = new JSObject();
                ret.put("success", true);
                call.resolve(ret);
            }

            @Override
            public void onError(String error) {
                call.reject(error);
            }
        });
    }

    @PluginMethod
    public void getPrinterStatus(PluginCall call) {
        String identifier = call.getString("identifier");
        if (identifier == null) {
            call.reject("Missing identifier parameter");
            return;
        }

        implementation.getPrinterStatus(identifier, new Zyprint.StatusCallback() {
            @Override
            public void onStatus(String status, String paperStatus, boolean connected) {
                JSObject ret = new JSObject();
                ret.put("status", status);
                ret.put("paperStatus", paperStatus);
                ret.put("connected", connected);
                call.resolve(ret);
            }

            @Override
            public void onError(String error) {
                call.reject(error);
            }
        });
    }
}
